### ISSUE TITLE
KongPlugin Config Does Not Support null/Absence of Values

### ISSUE THREAD

**Issue by iAnomaly** [Kong/kubernetes-ingress-controller#117]
### Summary

There is no declarative way of deleting an optional plugin configuration key via annotations (at least for CORS plugin).

For example, removing the headers config key annotation of the CORS plugin will NOT delete the key/values for that plugin in Kong. Further, using valid YAML/JSON values of null or empty string result in unexpected behavior: the CORS headers key for example results in a translated value of "*" rather than the deletion of the headers key entirely.

**Kong Ingress controller version**
0.0.4

### Kubernetes version

**paste `kubectl version` output**
v1.10.3

### What happened
The configuration key remains

### Expected behvaior
The configuration key should be removed from the plugin state in Kong

### Steps To Reproduce

1. Create a CORS plugin annotation with at least one optional key (such as headers) with value(s) of your choosing
2. Apply the annotation
3. Delete the headers key
4. Apply the updated annotation

### Painful Workaround
1. Delete all references to the plugin
2. Wait for Ingress controller to detect the change and remove the plugin
3. Remove the key you wish the delete from the plugin annotation
4. Restore the reference to the plugin

**Comment by iAnomaly** [comment]
It appears that plugins added to Kubernetes Service annotations suffer the same behavior: commenting/removing a plugin annotation such as `rate-limiting.plugin.konghq.com: whatever` does NOT trigger the Ingress controller to delete that plugin from Kong.

**Comment by jdevalk2** [comment]
I found the Kong Ingress supports create scenarios only for now.
The update code is not there yet.

The way to do this is delete the resources from kubernetes, and then re-add them.

I ran into a similar issue with creating ingress rules.

**Comment by hbagdi** [comment]
The upcoming 0.2.0 version of Ingress Controller will solve some of these issues around updating/deleting.

@iAnomaly It seems like `headers` property in CORS plugin is an array. Did you try setting `headers` to `[]`?

**Comment by iAnomaly** [comment]
Thanks @hbagdi. headers: [] does NOT work; Kong appears to treat it as NO headers rather than ANY headers. Guess we'll check out 0.2.0 when it comes out. Thanks.

**Comment by hbagdi** [comment]
@iAnomaly 0.2.0 was released just a while ago:
https://discuss.konghq.com/t/kong-ingress-controller-0-2-0-released/1921

**Comment by iAnomaly** [comment]
@hbagdi Just saw that! Upgraded to 0.2.0 and Kong 0.14 and tried it but still seeing the same behavior. More details below:

- Removing `headers` or using `headers: null`, ``headers: false`` > previous value persists/unchanged in Kong/database
- Use `headers: []`, `headers: ''` > `"headers": {}` in Kong/database results in NO headers returned rather than ANY requested header being returned (echoed/mirrored)

Appears there is still no way to get the Ingress Controller to notice when the key is absent/deleted in Kubernetes KongPlugin spec and update Kong's API/state accordingly.

**Comment by hbagdi** [comment]
@iAnomaly A series of issues are present here.

1. It seems like CORS plugin behavior is [different](https://github.com/Kong/kong/blob/master/kong/plugins/cors/handler.lua#L88) for an empty array(`headers: []`) and an absent `headers. 

2.
> Appears there is still no way to get the Ingress Controller to notice when the key is absent/deleted in Kubernetes KongPlugin spec and update Kong's API/state accordingly.

The underlying problem here is how to detect if a key present in a plugin config in Kong's DB  is a default key or was it set by a key in KongPlugin spec.

Allow me to explain:
Let's say we have a config:

```yaml
apiVersion: configuration.konghq.com/v1
kind: KongPlugin
metadata:
  name: rl
  lables:
config:
  second: 10
  redis_port: 4321
```

which creates the following plugin in Kong:

```json
{
            "config": {
                "fault_tolerant": true,
                "hide_client_headers": false,
                "hour": 102,
                "limit_by": "ip",
                "policy": "cluster",
                "redis_database": 0,
                "redis_port": 4321,
                "redis_timeout": 2000,
                "second": 10
            },
            "created_at": 1537561111000,
            "enabled": true,
            "id": "3bdc7a5e-a96a-4d87-84ee-882c49296229",
            "name": "rate-limiting",
            "service_id": "da9c4814-8411-4536-8d3f-b993835f6521"
        }
```

 Currently, Ingress Controller makes sures that all the key values present in KongPlugin spec match with the conf in Kong's DB.

Now, let's say you remove the `redis_port` field in KongPlugin spec.
There is no way to figure out if `redis_port` should be set to a value or set to default (by default it is set to 6379).

The work-around for this will be to use the [schema endpoint](https://docs.konghq.com/0.14.x/admin-api/#retrieve-plugin-schema) and then configure endpoints.

We will be addressing this issue in future. Thank you for the report!

**Comment by iAnomaly** [comment]
@hbagdi Thanks for explaining! Why not pass through the default behavior/validation to the upstream Kong API instead of deciding/enforcing in the Ingress Controller?

Using your example:
> Now, let's say you remove the redis_port field in KongPlugin spec.
There is no way to figure out if redis_port should be set to a value or set to default (by default it is set to 6379).

Don't try to figure out; just remove the `redis_port` field. My guess is for required fields the Kong API is going to error on the PATCH/POST and for non-required (optional) fields it will work as expected no? For failed/missing required fields the user will have to check the Ingress Controller logs for failures the same way they would any other client interacting with Kong's API.

I agree with you the better solution in the long run would be to use the schema endpoint to validate the configuration schema at Kubernetes apply time to fail at the KongPlugin resource layer instead of lower down from the Kong API. What do you think?

**Comment by hbagdi** [comment]
Hello @iAnomaly, what you suggest is probably not possible but I could be wrong.

> Don't try to figure out; just remove the redis_port field. My guess is for required fields the Kong API is going to error on the PATCH/POST and for non-required (optional) fields it will work as expected no? For failed/missing required fields the user will have to check the Ingress Controller logs for failures the same way they would any other client interacting with Kong's API.

Kong Ingress Controller syncs up entities in Kong with resources in Kubernetes.
It does so by running a [reconciliation](https://hackernoon.com/level-triggering-and-reconciliation-in-kubernetes-1f17fe30333d) loop to make sure that configuration of Kong is up-to-date with the desired configuration (the one in Kubernetes).
Ingress Controller diffs the two entities (in k8s and Kong) and then issues a POST/PUT/PATCH/DELETE when needed.
We avoid directly firing these update requests all the time since that will lead to reduced performance on the Kong side and most of the time, it will be unnecessary any ways. Hence, it is important to be able to diff and then apply the changes when needed.

In future, it will be possible to consider defaults while doing diff comparisons but unfortunately it's not possible today.

**Comment by hbagdi** [comment]
Update: 
With Kong 1.0, we get `PUT` support on Kong's Admin API for plugin configuration, allowing us to be explicit about the updates we wish to make to plugin objects and override default values.

this will be addressed as we work towards #241.

**Comment by iAnomaly** [comment]
Awesome! 👏 Thank you for the update @hbagdi. 🙇

**Comment by hbagdi** [comment]
This is now fixed with 0.4.0.
